src/vos/vos_dtx.c

```diff
@@ -1369,8 +1383,19 @@ vos_dtx_commit_internal(struct vos_container *cont, struct dtx_id *dtis,
 
 	if (slots > 1) {
 		D_ALLOC(dce_df, sizeof(*dce_df) * slots);
-		if (dce_df == NULL)
-			return -DER_NOMEM;
+		if (dce_df == NULL) {
+			D_ERROR("Not enough DRAM to commit "DF_DTI"\n",
+				DP_DTI(&dtis[cur]));
+
+			/* For the DTXs that have been committed we will not
+			 * re-insert them back into the active DTX table (in
+			 * DRAM) even if we abort the PMDK transaction, then
+			 * let's hide the error and commit former successful
+			 * DTXs. The left non-committed DTXs will be handled
+			 * next time.
+			 */
```

Comment:
FYI, to deal with the problems of rollback DRAM changes on transaction abort, you can make the DRAM changes in commit callback. See umem_tx_add_callback().

I don't think your patch can solve the problem completely, because we can't avoid transaction abort in some cases. (what if pmemobj calls failed and caused transaction abort?)

Reply:
The mainly issue is not about abort the PKDM transition, instead, it is about how to rollback the in-DRAM changes of the DTX index: active DTX table and committed DTX table, that is out of PMDK control when abort the PMDK transaction.

Reply:
what happens with the in-memory structures if the transaction fails to commit ?

Reply:
That is good question. If the PMDK transaction failed to be committed, but we have already modified the in-DRAM btree (DTX table), then we may have to restart the server to re-establish the DTX index from the SCM. Anyway, that is out of current patch scope.